In [1]:
import pandas as pd, numpy as np
from matplotlib import pyplot as plt

import scipy.stats  as stats

In [2]:
pd.options.display.max_columns = 50

In [4]:
best = pd.read_csv("../v08000/submit/ensamble_0.50633.csv.gz", compression='gzip')
best.head()

id        F1        F2        F3        F4  \
0  HOBBIES_1_001_CA_1_validation  0.763760  0.903445  0.936408  1.176313   
1  HOBBIES_1_002_CA_1_validation  0.167927  0.136299  0.175232  0.223089   
2  HOBBIES_1_003_CA_1_validation  0.248042  0.339250  0.391313  0.334560   
3  HOBBIES_1_004_CA_1_validation  1.526960  1.493865  1.488617  1.564704   
4  HOBBIES_1_005_CA_1_validation  0.878413  0.657317  0.903145  0.855637   

         F5        F6        F7        F8        F9       F10       F11  \
0  0.904619  1.771964  1.985657  0.566341  0.568393  0.607699  0.440619   
1  0.208410  0.233510  0.276279  0.294497  0.454775  0.413967  0.321073   
2  0.440682  0.506670  0.575179  0.293484  0.299479  0.250709  0.445623   
3  1.773607  2.277775  3.404197  1.659009  1.493923  1.370968  1.782565   
4  0.906150  1.167033  1.319796  1.243085  1.068527  0.949900  0.890768   

        F12       F13       F14       F15       F16       F17       F18  \
0  0.638120  1.108987  0.665889  0.788719  0.591017  0.604705  0.709199   
1  0.458137  0.350087  0.404353  0.237395  0.277815  0.283098  0.256969   
2  0.340701  0.883335  0.620698  0.349982  0.424201  0.416743  0.444411   
3  2.069228  3.376062  3.085069  1.410725  1.124932  1.162898  1.229955   
4  0.989954  1.804170  1.360495  0.644089  0.652468  0.675208  0.681183   

        F19       F20       F21       F22       F23       F24       F25  \
0  0.698358  1.482235  0.708192  0.594468  0.540233  0.453303  0.592453   
1  0.184456  0.403093  0.394788  0.256325  0.255236  0.219102  0.341571   
2  0.669945  0.627456  0.691125  0.408742  0.312358  0.406725  0.616306   
3  1.217408  2.191248  2.203557  1.945107  1.479941  1.541212  1.357651   
4  0.742110  0.961295  0.943992  1.252646  1.063600  1.314395  1.303209   

        F26       F27       F28  
0  0.691477  0.748558  0.830309  
1  0.302225  0.356841  0.392271  
2  0.625662  0.536575  0.812369  
3  2.068405  2.962143  4.015549  
4  1.451894  2.407974  2.176170

In [5]:
sales = pd.read_csv("../data/raw/sales_train_validation.csv")
sales.head()

id        item_id    dept_id   cat_id store_id  \
0  HOBBIES_1_001_CA_1_validation  HOBBIES_1_001  HOBBIES_1  HOBBIES     CA_1   
1  HOBBIES_1_002_CA_1_validation  HOBBIES_1_002  HOBBIES_1  HOBBIES     CA_1   
2  HOBBIES_1_003_CA_1_validation  HOBBIES_1_003  HOBBIES_1  HOBBIES     CA_1   
3  HOBBIES_1_004_CA_1_validation  HOBBIES_1_004  HOBBIES_1  HOBBIES     CA_1   
4  HOBBIES_1_005_CA_1_validation  HOBBIES_1_005  HOBBIES_1  HOBBIES     CA_1   

  state_id  d_1  d_2  d_3  d_4  d_5  d_6  d_7  d_8  d_9  d_10  d_11  d_12  \
0       CA    0    0    0    0    0    0    0    0    0     0     0     0   
1       CA    0    0    0    0    0    0    0    0    0     0     0     0   
2       CA    0    0    0    0    0    0    0    0    0     0     0     0   
3       CA    0    0    0    0    0    0    0    0    0     0     0     0   
4       CA    0    0    0    0    0    0    0    0    0     0     0     0   

   d_13  d_14  d_15  d_16  d_17  d_18  d_19  ...  d_1889  d_1890  d_1891  \
0     0     0     0     0     0     0     0  ...       0       0       0   
1     0     0     0     0     0     0     0  ...       0       0       0   
2     0     0     0     0     0     0     0  ...       0       0       0   
3     0     0     0     0     0     0     0  ...       0       3       1   
4     0     0     0     0     0     0     0  ...       4       0       1   

   d_1892  d_1893  d_1894  d_1895  d_1896  d_1897  d_1898  d_1899  d_1900  \
0       1       0       4       2       3       0       1       2       0   
1       0       0       0       0       0       0       0       0       0   
2       1       0       0       0       1       0       0       0       0   
3       2       1       3       1       0       2       5       4       2   
4       4       0       1       0       1       0       1       1       2   

   d_1901  d_1902  d_1903  d_1904  d_1905  d_1906  d_1907  d_1908  d_1909  \
0       0       0       1       1       3       0       1       1       1   
1       0       0       0       0       0       0       0       0       1   
2       0       1       2       2       1       2       1       1       1   
3       0       3       0       1       0       5       4       1       0   
4       0       1       1       2       1       1       0       1       1   

   d_1910  d_1911  d_1912  d_1913  
0       3       0       1       1  
1       0       0       0       0  
2       0       1       1       1  
3       1       3       7       2  
4       2       2       2       4  

[5 rows x 1919 columns]

In [6]:
sub = best.merge(sales[["id", "item_id", "dept_id", "cat_id", "store_id", "state_id"]], on = "id")
sub["_all_"] = "Total"
sub.shape

(30490, 35)

In [7]:
sub.head()

id        F1        F2        F3        F4  \
0  HOBBIES_1_001_CA_1_validation  0.763760  0.903445  0.936408  1.176313   
1  HOBBIES_1_002_CA_1_validation  0.167927  0.136299  0.175232  0.223089   
2  HOBBIES_1_003_CA_1_validation  0.248042  0.339250  0.391313  0.334560   
3  HOBBIES_1_004_CA_1_validation  1.526960  1.493865  1.488617  1.564704   
4  HOBBIES_1_005_CA_1_validation  0.878413  0.657317  0.903145  0.855637   

         F5        F6        F7        F8        F9       F10       F11  \
0  0.904619  1.771964  1.985657  0.566341  0.568393  0.607699  0.440619   
1  0.208410  0.233510  0.276279  0.294497  0.454775  0.413967  0.321073   
2  0.440682  0.506670  0.575179  0.293484  0.299479  0.250709  0.445623   
3  1.773607  2.277775  3.404197  1.659009  1.493923  1.370968  1.782565   
4  0.906150  1.167033  1.319796  1.243085  1.068527  0.949900  0.890768   

        F12       F13       F14       F15       F16       F17       F18  \
0  0.638120  1.108987  0.665889  0.788719  0.591017  0.604705  0.709199   
1  0.458137  0.350087  0.404353  0.237395  0.277815  0.283098  0.256969   
2  0.340701  0.883335  0.620698  0.349982  0.424201  0.416743  0.444411   
3  2.069228  3.376062  3.085069  1.410725  1.124932  1.162898  1.229955   
4  0.989954  1.804170  1.360495  0.644089  0.652468  0.675208  0.681183   

        F19       F20       F21       F22       F23       F24       F25  \
0  0.698358  1.482235  0.708192  0.594468  0.540233  0.453303  0.592453   
1  0.184456  0.403093  0.394788  0.256325  0.255236  0.219102  0.341571   
2  0.669945  0.627456  0.691125  0.408742  0.312358  0.406725  0.616306   
3  1.217408  2.191248  2.203557  1.945107  1.479941  1.541212  1.357651   
4  0.742110  0.961295  0.943992  1.252646  1.063600  1.314395  1.303209   

        F26       F27       F28        item_id    dept_id   cat_id store_id  \
0  0.691477  0.748558  0.830309  HOBBIES_1_001  HOBBIES_1  HOBBIES     CA_1   
1  0.302225  0.356841  0.392271  HOBBIES_1_002  HOBBIES_1  HOBBIES     CA_1   
2  0.625662  0.536575  0.812369  HOBBIES_1_003  HOBBIES_1  HOBBIES     CA_1   
3  2.068405  2.962143  4.015549  HOBBIES_1_004  HOBBIES_1  HOBBIES     CA_1   
4  1.451894  2.407974  2.176170  HOBBIES_1_005  HOBBIES_1  HOBBIES     CA_1   

  state_id  _all_  
0       CA  Total  
1       CA  Total  
2       CA  Total  
3       CA  Total  
4       CA  Total

## Different ratios for different aggregation levels

The higher the aggregation level, the more confident we are in the point prediction --> lower coef, relatively smaller range of quantiles

In [8]:
qs = np.array([0.005,0.025,0.165,0.25, 0.5, 0.75, 0.835, 0.975, 0.995])
qs.shape


def get_ratios(coef=0.15):
    qs2 = np.log(qs/(1-qs))*coef
    ratios = stats.norm.cdf(qs2)
    ratios /= ratios[4]
    ratios = pd.Series(ratios, index=qs)
    return ratios.round(3)

#coef between 0.05 and 0.24 is used, probably suboptimal values for now

level_coef_dict = {"id": get_ratios(coef=0.3), "item_id": get_ratios(coef=0.15),
                   "dept_id": get_ratios(coef=0.08), "cat_id": get_ratios(coef=0.07),
                   "store_id": get_ratios(coef=0.08), "state_id": get_ratios(coef=0.07), "_all_": get_ratios(coef=0.05),
                   ("state_id", "item_id"): get_ratios(coef=0.19),  ("state_id", "dept_id"): get_ratios(coef=0.1),
                    ("store_id","dept_id") : get_ratios(coef=0.11), ("state_id", "cat_id"): get_ratios(coef=0.08),
                    ("store_id","cat_id"): get_ratios(coef=0.1)
                  }



Let's see how ranges differ!

In [9]:
level_coef_dict["id"]

0.005    0.112
0.025    0.272
0.165    0.627
0.250    0.742
0.500    1.000
0.750    1.258
0.835    1.373
0.975    1.728
0.995    1.888
dtype: float64

In [10]:
level_coef_dict["cat_id"]

0.005    0.711
0.025    0.798
0.165    0.910
0.250    0.939
0.500    1.000
0.750    1.061
0.835    1.090
0.975    1.202
0.995    1.289
dtype: float64

For the the lowest level (30490 series), the smallest and biggest quantiles are 20% and 180% of the point prediction. For categories (3 series), the model will be way more confident: the smallest quantile will be 71%, the biggest will be 129% of the point prediction.

In [11]:
def quantile_coefs(q, level):
    ratios = level_coef_dict[level]
               
    return ratios.loc[q].values

def get_group_preds(pred, level):
    df = pred.groupby(level)[cols].sum()
    q = np.repeat(qs, len(df))
    df = pd.concat([df]*9, axis=0, sort=False)
    df.reset_index(inplace = True)
    df[cols] *= quantile_coefs(q, level)[:, None]
    if level != "id":
        df["id"] = [f"{lev}_X_{q:.3f}_validation" for lev, q in zip(df[level].values, q)]
    else:
        df["id"] = [f"{lev.replace('_validation', '')}_{q:.3f}_validation" for lev, q in zip(df[level].values, q)]
    df = df[["id"]+list(cols)]
    return df

def get_couple_group_preds(pred, level1, level2):
    df = pred.groupby([level1, level2])[cols].sum()
    q = np.repeat(qs, len(df))
    df = pd.concat([df]*9, axis=0, sort=False)
    df.reset_index(inplace = True)
    df[cols] *= quantile_coefs(q, (level1, level2))[:, None]
    df["id"] = [f"{lev1}_{lev2}_{q:.3f}_validation" for lev1,lev2, q in 
                zip(df[level1].values,df[level2].values, q)]
    df = df[["id"]+list(cols)]
    return df

levels = ["id", "item_id", "dept_id", "cat_id", "store_id", "state_id", "_all_"]
couples = [("state_id", "item_id"),  ("state_id", "dept_id"),("store_id","dept_id"),
                            ("state_id", "cat_id"),("store_id","cat_id")]
cols = [f"F{i}" for i in range(1, 29)]

df = []
for level in levels :
    df.append(get_group_preds(sub, level))
for level1,level2 in couples:
    df.append(get_couple_group_preds(sub, level1, level2))
df = pd.concat(df, axis=0, sort=False)
df.reset_index(drop=True, inplace=True)
df = pd.concat([df,df] , axis=0, sort=False)
df.reset_index(drop=True, inplace=True)
df.loc[df.index >= len(df.index)//2, "id"] = df.loc[df.index >= len(df.index)//2, "id"].str.replace("_validation$", "_evaluation")

In [12]:
print(df.shape)
df.head()

(771120, 29)


id        F1        F2        F3        F4  \
0  FOODS_1_001_CA_1_0.005_validation  0.079706  0.071019  0.070553  0.070608   
1  FOODS_1_001_CA_2_0.005_validation  0.102895  0.109957  0.095857  0.107499   
2  FOODS_1_001_CA_3_0.005_validation  0.120851  0.122073  0.094418  0.112105   
3  FOODS_1_001_CA_4_0.005_validation  0.055167  0.045611  0.055074  0.059578   
4  FOODS_1_001_TX_1_0.005_validation  0.063586  0.069404  0.060041  0.062800   

         F5        F6        F7        F8        F9       F10       F11  \
0  0.078404  0.094313  0.089551  0.137959  0.116186  0.127685  0.107127   
1  0.124137  0.129265  0.135279  0.097528  0.101935  0.093784  0.099949   
2  0.146573  0.117336  0.127321  0.102197  0.128224  0.114178  0.127227   
3  0.050038  0.041145  0.065198  0.051901  0.047152  0.042680  0.041904   
4  0.059490  0.077878  0.066676  0.072656  0.078819  0.072224  0.067293   

        F12       F13       F14       F15       F16       F17       F18  \
0  0.148987  0.115104  0.143923  0.096864  0.103392  0.105963  0.087865   
1  0.097087  0.116042  0.111412  0.090190  0.085957  0.092493  0.083711   
2  0.144831  0.156598  0.159673  0.102853  0.114885  0.105647  0.095973   
3  0.033335  0.056651  0.056125  0.055500  0.049123  0.055439  0.062269   
4  0.073442  0.075259  0.075258  0.052515  0.054048  0.054188  0.061809   

        F19       F20       F21       F22       F23       F24       F25  \
0  0.126670  0.125062  0.103461  0.090613  0.076115  0.091398  0.078016   
1  0.099713  0.144655  0.145430  0.101650  0.122791  0.120085  0.134281   
2  0.142586  0.186673  0.209681  0.127257  0.102255  0.140614  0.110007   
3  0.042856  0.054620  0.057459  0.030558  0.037584  0.036504  0.037157   
4  0.050225  0.070437  0.068553  0.083766  0.095978  0.077748  0.105620   

        F26       F27       F28  
0  0.094548  0.101360  0.093343  
1  0.132484  0.180562  0.171010  
2  0.153905  0.119478  0.157536  
3  0.035720  0.041661  0.042108  
4  0.078122  0.115187  0.089891

In [13]:
df.to_csv("../data/submit/ensemble_point_to_uncertainty.csv.gz", index=False, compression='gzip')